src/internal/tester/its/m365.go

```diff
@@ -0,0 +1,209 @@
+package its
+
+import (
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/alcionai/clues"
+	"github.com/h2non/gock"
+	"github.com/stretchr/testify/require"
+
+	"github.com/alcionai/corso/src/internal/common/idname"
+	"github.com/alcionai/corso/src/internal/common/ptr"
+	"github.com/alcionai/corso/src/internal/tester"
+	"github.com/alcionai/corso/src/internal/tester/tconfig"
+	"github.com/alcionai/corso/src/pkg/account"
+	"github.com/alcionai/corso/src/pkg/control"
+	"github.com/alcionai/corso/src/pkg/count"
+	"github.com/alcionai/corso/src/pkg/services/m365/api"
+	"github.com/alcionai/corso/src/pkg/services/m365/api/graph"
+)
+
+// ---------------------------------------------------------------------------
+// Gockable client
+// ---------------------------------------------------------------------------
+
+// GockClient produces a new exchange api client that can be
+// mocked using gock.
+func gockClient(creds account.M365Config, counter *count.Bus) (api.Client, error) {
+	s, err := graph.NewGockService(creds, counter)
+	if err != nil {
+		return api.Client{}, err
+	}
+
+	li, err := graph.NewGockService(creds, counter, graph.NoTimeout())
+	if err != nil {
+		return api.Client{}, err
+	}
+
+	return api.Client{
+		Credentials: creds,
+		Stable:      s,
+		LargeItem:   li,
+	}, nil
+}
+
+// ---------------------------------------------------------------------------
+// Intercepting calls with Gock
+// ---------------------------------------------------------------------------
+
+const graphAPIHostURL = "https://graph.microsoft.com"
+
+func V1APIURLPath(parts ...string) string {
+	return strings.Join(append([]string{"/v1.0"}, parts...), "/")
+}
+
+func InterceptV1Path(pathParts ...string) *gock.Request {
+	return gock.New(graphAPIHostURL).Get(V1APIURLPath(pathParts...))
+}
+
+// ---------------------------------------------------------------------------
+// Suite Setup
+// ---------------------------------------------------------------------------
+
+type IDs struct {
+	Provider          idname.Provider
+	ID                string
+	Email             string
+	DriveID           string
+	DriveRootFolderID string
+	TestContainerID   string
+	WebURL            string
+}
+
+// m365IntgTestSetup provides all the common references used in an m365 integration
+// test suite.  Call `its.GetM365()` to get the singleton for your test suite.
+// If you're looking for unit testing setup, use `uts.GetM365()` instead.
+type m365IntgTestSetup struct {
+	Acct     account.Account
+	Creds    account.M365Config
+	TenantID string
+
+	AC     api.Client
+	GockAC api.Client
+
+	Site  IDs
+	Group IDs
+
+	User          IDs
+	SecondaryUser IDs
+	TertiaryUser  IDs
+}
+
+var (
+	singleton *m365IntgTestSetup
+	mu        sync.Mutex
+)
+
+// GetM365 returns the populated its.m365 singleton.
+func GetM365(t *testing.T) m365IntgTestSetup {
+	mu.Lock()
+	defer mu.Unlock()
+
+	if singleton != nil {
+		return *singleton
+	}
```

Comment:
is there a particular need for this to be a singleton? Often singletons cause more problems than the solve because it's not always apparent when they're initialized or what they're initialized with. The global concurrency limiter we have is an excellent example of why singletons can be confusing/error prone to use

Reply:
It's not a necessity, no.  But I'd say this is a fairly safe case since the state of initialization is resolved at the time of accessing the singleton (which is non-optional for usage), and there's no difference in what data is produced since that's all controlled internal to the package.  Our other singletons are more problematic because initialization is optional, and singletons can be initialized with different configurations, neither of which is the case here.

Reply:
I think whether it's a safe case or not depends on the context of how it's used. The next PR makes this an exported struct, so we now have an exported singleton where many of it's fields are also exported. If any test decides that the way to pass around non-standard data to later parts of a test is to modify the fields in the struct then all tests could be affected. Thus, this ends up encoding project level standards in places where they may not really be required

Reply:
That should be safe to do, since the getter returns a copy of the singleton, not its reference.

Reply:
that only really works assuming everything remains basic types instead of more complex things like interfaces, pointers, maps, slices, etc. The fact that we return a copy to this type really makes we wonder why it's a singleton though. It seems like it would be less complex to just make a new set of data all the time

Reply:
mostly trying to minimize the re-calling of existing data.  It's not a critical performance issue.

Reply:
I think avoiding repopulating data is a valid goal, but I feel this particular pattern is a little too error prone to outweigh the possible benefits. Although this is code for tests, our heavy reliance on automated testing means that we need to have high confidence that our tests are doing what we think they are even as things change over time. Since misuse or unintended consequences from changes to this code could affect a large chunk of the integration tests in the repo, we should make sure it's difficult to misuse or break with future changes

That said, there are other designs that would allow for fetching data only once in a safe manner which we can explore in parallel to this PR going forward (or after) if you want